lib/framework/MqttSettingsService.cpp

```diff
@@ -207,6 +207,17 @@ bool MqttSettingsService::configureMqtt() {
         static_cast<espMqttClient *>(_mqttClient)->setClientId(_state.clientId.c_str());
         static_cast<espMqttClient *>(_mqttClient)->setKeepAlive(_state.keepAlive);
         static_cast<espMqttClient *>(_mqttClient)->setCleanSession(_state.cleanSession);
+
+        // create last will topic with the base prefixed. It has to be static because the client destroys the reference
+        static char will_topic[FACTORY_MQTT_MAX_TOPIC_LENGTH];
+        if (_state.base.isEmpty()) {
+            snprintf(will_topic, sizeof(will_topic), "status");
+        } else {
+            snprintf(will_topic, sizeof(will_topic), "%s/status", _state.base.c_str());
+        }
+        setWill(will_topic);
+
+
```

Comment:
This only effects the unsecure client. 
Better remove `setWill()`function, set the will_topic before line 187 (#ifndef TASMOTA) and use 
`static_cast<espMqttClientSecure *>(_mqttClient)->setWill(will_topic, 1, true, "offline");`
and
`static_cast<espMqttClient *>(_mqttClient)->setWill(will_topic, 1, true, "offline");`
in the secure/unsecure sections (before line 200/210)

Reply:
setWill does before secure and unsecure, but indeed this approach is cleaner. I'll update